File: packages/android/src/main/java/com/kozmos/components/Sidebar/Sidebar.kt
package com.kozmos.components.sidebar

import com.kozmos.tokens.KozmosDimensions

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.PermanentDrawerSheet
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.kozmos.tokens.KozmosThemeTokens

enum class KozmosSidebarVariant {
    Expanded,
    Rail
}

@Composable
fun KozmosSidebar(
    modifier: Modifier = Modifier,
    content: @Composable ColumnScope.() -> Unit
) {
    KozmosSidebar(
        modifier = modifier,
        navigation = content
    )
}

@Composable
fun KozmosSidebar(
    modifier: Modifier = Modifier,
    variant: KozmosSidebarVariant = KozmosSidebarVariant.Expanded,
    header: @Composable ColumnScope.() -> Unit = {},
    navigation: @Composable ColumnScope.() -> Unit,
    tools: @Composable ColumnScope.() -> Unit = {},
    footer: @Composable ColumnScope.() -> Unit = {}
) {
    val isRail = variant == KozmosSidebarVariant.Rail
    PermanentDrawerSheet(
        modifier = modifier.width(if (isRail) 80.dp else 280.dp),
        drawerContainerColor = KozmosThemeTokens.primitivesColorsBackground0,
    ) {
        Column(
            modifier = Modifier
                .padding(
                    horizontal = if (isRail) KozmosDimensions.primitivesLayoutSpacing100 else KozmosDimensions.primitivesLayoutSpacing200,
                    vertical = KozmosDimensions.primitivesLayoutSpacing300
                )
                .fillMaxHeight(),
            horizontalAlignment = if (isRail) Alignment.CenterHorizontally else Alignment.Start,
            verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing200)
        ) {
            header()
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .weight(1f),
                horizontalAlignment = if (isRail) Alignment.CenterHorizontally else Alignment.Start,
                verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing100),
                content = navigation
            )
            tools()
            footer()
        }
    }
}

File: packages/ios/Sources/Components/Sidebar/Sidebar.swift
import SwiftUI

public enum KozmosSidebarVariant {
    case expanded
    case rail
}

public struct KozmosSidebar<Content: View>: View {
    let variant: KozmosSidebarVariant
    let header: AnyView?
    let content: Content
    let tools: AnyView?
    let footer: AnyView?

    public init(
        variant: KozmosSidebarVariant = .expanded,
        @ViewBuilder content: () -> Content
    ) {
        self.variant = variant
        self.header = nil
        self.content = content()
        self.tools = nil
        self.footer = nil
    }

    public init<Header: View, Tools: View, Footer: View>(
        variant: KozmosSidebarVariant = .expanded,
        @ViewBuilder header: () -> Header,
        @ViewBuilder navigation: () -> Content,
        @ViewBuilder tools: () -> Tools,
        @ViewBuilder footer: () -> Footer
    ) {
        self.variant = variant
        self.header = AnyView(header())
        self.content = navigation()
        self.tools = AnyView(tools())
        self.footer = AnyView(footer())
    }

    public var body: some View {
        HStack(spacing: KozmosDimensions.primitivesLayoutSpacing0) {
            VStack(alignment: stackAlignment, spacing: KozmosDimensions.primitivesLayoutSpacing200) {
                if let header = header {
                    header
                        .frame(maxWidth: .infinity, alignment: contentAlignment)
                }

                ScrollView {
                    content
                        .frame(maxWidth: .infinity, alignment: contentAlignment)
                }
                .frame(maxHeight: .infinity)

                if let tools = tools {
                    tools
                        .frame(maxWidth: .infinity, alignment: contentAlignment)
                }

                if let footer = footer {
                    footer
                        .frame(maxWidth: .infinity, alignment: contentAlignment)
                }
            }
            .padding(.horizontal, horizontalPadding)
            .padding(.vertical, KozmosDimensions.primitivesLayoutSpacing300)
            .frame(width: width)
            .frame(maxHeight: .infinity)
            .background(KozmosColors.primitivesColorsBackground0)
            .overlay(
                Rectangle()
                    .frame(width: 1)
                    .foregroundColor(KozmosColors.primitivesColorsBackground300),
                alignment: .trailing
            )

            Spacer()
        }
    }

    private var width: CGFloat {
        variant == .rail ? 80 : 256
    }

    private var horizontalPadding: CGFloat {
        variant == .rail ? KozmosDimensions.primitivesLayoutSpacing100 : KozmosDimensions.primitivesLayoutSpacing200
    }

    private var stackAlignment: HorizontalAlignment {
        variant == .rail ? .center : .leading
    }

    private var contentAlignment: Alignment {
        variant == .rail ? .center : .leading
    }
}
